File: src/domain/personalization/ProgressEngine.ts
import { SkillScore } from '../entities/SkillScore';
import { LearningPathNode } from '../entities/LearningPathNode';
import { REINFORCEMENT_XP_REWARD, LESSON_XP_REWARD } from '@/core/constants/adaptiveThresholds';

export class ProgressEngine {
  /**
   * Calculates overall course mastery % based on average skill scores
   */
  public static calculateCourseMastery(skillScores: Record<string, SkillScore>): number {
    const scores = Object.values(skillScores);
    if (scores.length === 0) return 0;
    const total = scores.reduce((sum, item) => sum + item.score, 0);
    return Math.round(total / scores.length);
  }

  /**
   * Calculates roadmap completion percentage
   */
  public static calculateRoadmapProgress(nodes: LearningPathNode[]): number {
    if (nodes.length === 0) return 0;
    const completedCount = nodes.filter((n) => n.status === 'completed').length;
    return Math.round((completedCount / nodes.length) * 100);
  }

  /**
   * Safe idempotent XP adder preventing double-awarding
   */
  public static calculateNewXP(
    currentXP: number,
    actionType: 'lesson_completion' | 'reinforcement_completion',
    actionId: string,
    completedActionIds: string[]
  ): { newXP: number; wasAwarded: boolean } {
    if (completedActionIds.includes(actionId)) {
      return { newXP: currentXP, wasAwarded: false };
    }

    const reward = actionType === 'reinforcement_completion' ? REINFORCEMENT_XP_REWARD : LESSON_XP_REWARD;
    return {
      newXP: currentXP + reward,
      wasAwarded: true,
    };
  }
}

File: mobile/lib/domain/personalization/progress_engine.dart
import '../entities/skill_score.dart';
import '../entities/learning_path_node.dart';
import '../../core/constants/adaptive_thresholds.dart';

class ProgressEngine {
  /// Calculates overall course mastery % based on average skill scores
  static int calculateCourseMastery(Map<String, SkillScore> skillScores) {
    final scores = skillScores.values.toList();
    if (scores.isEmpty) return 0;
    final total = scores.fold<int>(0, (sum, item) => sum + item.score);
    return (total / scores.length).round();
  }

  /// Calculates roadmap completion percentage
  static int calculateRoadmapProgress(List<LearningPathNode> nodes) {
    if (nodes.isEmpty) return 0;
    final completedCount = nodes.where((n) => n.status == NodeStatus.completed).length;
    return ((completedCount / nodes.length) * 100).round();
  }

  /// Safe idempotent XP adder preventing double-awarding
  static ({int newXP, bool wasAwarded}) calculateNewXP({
    required int currentXP,
    required String actionType, // 'lesson_completion' | 'reinforcement_completion'
    required String actionId,
    required List<String> completedActionIds,
  }) {
    if (completedActionIds.contains(actionId)) {
      return (newXP: currentXP, wasAwarded: false);
    }

    final reward = actionType == 'reinforcement_completion'
        ? AdaptiveThresholds.reinforcementXpReward
        : AdaptiveThresholds.lessonXpReward;

    return (
      newXP: currentXP + reward,
      wasAwarded: true,
    );
  }
}
